Add api.content.iter_ancestors (#562)
plone/plone.api: docs/content.md

%
% self.assertEqual(view.__name__, u'plone')

(content-get-path-example)=

## Get content path

To get the path of a content object, use {func}`api.content.get_path`.
This function accepts an object for which you want to get its path as the required parameter `obj`, and an optional boolean parameter `relative` whose default is `False`.

It returns either an absolute path from the Zope root by default or when `relative` is set to `False`, or a relative path from the portal root when `relative` is set to `True`.

The following example shows how to get the absolute path from the Zope root.

```python
from plone import api
portal = api.portal.get()

> folder = portal['events']['training']

Reviewer: ```suggestion
folder = portal["events"]["training"]
```

path = api.content.get_path(obj=folder)

> assert path == '/plone/events/training'

Reviewer: ```suggestion
assert path == "/plone/events/training"
```

```

The following example shows how to get the portal-relative path.

```python
rel_path = api.content.get_path(obj=folder, relative=True)

> assert rel_path == 'events/training'

Reviewer: ```suggestion
assert rel_path == "events/training"
```

```

If the API is used to fetch an object with the `relative` parameter set as `True`, and the object is outside the portal, it throws an `InvalidParameterError` error.

% invisible-code-block: python
%
% # Setup an object outside portal for testing error case
% app = portal.aq_parent

> % app.manage_addFolder('outside_folder')

Reviewer: ```suggestion
% app.manage_addFolder("outside_folder")
```

%
% # Test that getting relative path for object outside portal raises error
% from plone.api.exc import InvalidParameterError
% with self.assertRaises(InvalidParameterError):
%     api.content.get_path(obj=app.outside_folder, relative=True)

```python
from plone.api.exc import InvalidParameterError

# Getting path of an object outside portal raises InvalidParameterError
try:
    outside_path = api.content.get_path(
        obj=app.outside_folder,
        relative=True
    )
    assert False, "Should raise InvalidParameterError and not reach this code"
except InvalidParameterError as e:
    assert "Object not in portal path" in str(e)
```

(content-iter-ancestors-example)=

## Iterate over object ancestors

> To iterate over the ancestors in an object acquisition chain, use the {func}`api.content.iter_ancestors` function.

Reviewer 2: ```suggestion
To iterate over the ancestors in the object tree, use the {func}`api.content.iter_ancestors` function.
```

Reviewer: Why not have a convenience method of `get_closest_parent` that gets the first iteration of `api.content.iter_ancestors` with either `[0]` or `next`? I'm not sure if it's performant with a large or deep hierarchy.

```python
from plone import api
portal = api.portal.get()

# Get all ancestors of the team object
ancestors = api.content.iter_ancestors(portal.about.team)
```

% invisible-code-block: python
%
% self.assertTupleEqual(tuple(ancestors), (portal.about, portal))

To iterate over the ancestors that implement one interface, use the {func}`api.content.iter_ancestors` function with the `interface` argument.

```python
from plone import api
from Products.CMFCore.interfaces import ISiteRoot
portal = api.portal.get()

# Get all ancestors of the team object that implement the ISiteRoot interface
ancestors = api.content.iter_ancestors(portal.about.team, interface=ISiteRoot)
```

% invisible-code-block: python
%
% self.assertTupleEqual(tuple(ancestors), (portal,))

To iterate over the ancestors using a custom filter you can use the {func}`api.content.iter_ancestors` function with the `function` argument.

```python
from plone import api
portal = api.portal.get()

# Get all ancestors of the team object with the id 'about'

> ancestors = api.content.iter_ancestors(portal.about.team, function=lambda obj: obj.id == 'about')

Reviewer: ```suggestion
ancestors = api.content.iter_ancestors(portal.about.team, function=lambda obj: obj.id == "about")
```

```

% invisible-code-block: python
%
% self.assertTupleEqual(tuple(ancestors), (portal.about,))

> To iterate over the ancestors until an object is found you use the {func}`api.content.iter_ancestors` function with the `stop_at` argument.

Reviewer: ```suggestion
To iterate over the ancestors until an object is found, use the {func}`api.content.iter_ancestors` function with the `stop_at` argument.
```

%
% self.assertTupleEqual(tuple(ancestors), (portal.about,))

(content-get-closest-ancestor-example)=

## Get closest ancestor

To get the closest ancestor in the object tree, use the {func}`api.content.get_closest_ancestor` function.

```python
from plone import api
from Products.CMFCore.interfaces import ISiteRoot
portal = api.portal.get()

# Get the closest ancestor of the team object that implements the ISiteRoot interface
ancestor = api.content.get_closest_ancestor(portal.about.team, interface=ISiteRoot)
```

% invisible-code-block: python
%
% self.assertEqual(ancestor, portal)

> To get the closest ancestor that using a custom filter, use the {func}`api.content.get_closest_ancestor` function with the `function` argument.

Reviewer: ```suggestion
To get the closest ancestor by using a custom filter, use the {func}`api.content.get_closest_ancestor` function with the `function` argument.
```
